pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -1143,15 +1142,35 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 	return tlsContext, nil
 }
 
-func (cb *ClusterBuilder) setAutoSni(mc *MutableCluster, tls *networking.ClientTLSSettings) {
-	if mc != nil && features.EnableAutoSni && len(tls.Sni) == 0 {
+// Set auto_sni if EnableAutoSni feature flag is enabled and if sni field is not explicitly set in DR.
+// Set auto_san_validation if VerifyCertAtClient feature flag is enabled and if there is no explicit SubjectAltNames specified  in DR.
+func (cb *ClusterBuilder) setAutoSniAndAutoSanValidation(mc *MutableCluster, tls *networking.ClientTLSSettings) {
+	if mc == nil || !features.EnableAutoSni {
+		return
+	}
+
+	setAutoSni := false
+	setAutoSanValidation := false
+	if len(tls.Sni) == 0 {
+		setAutoSni = true
+	}
+	if features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0 {
+		setAutoSanValidation = true
+	}
+
+	if setAutoSni || setAutoSanValidation {
 		if mc.httpProtocolOptions == nil {
 			mc.httpProtocolOptions = &http.HttpProtocolOptions{}
 		}
 		if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
 			mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
 		}
-		mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSni = true
+		if setAutoSni {
+			mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSni = true
+		}
+		if setAutoSanValidation {
```

Comment:
the auto san validation should be set with auto sni

Reply:
@hzxuzhonghu Please take a look at my [comment](https://github.com/istio/istio/pull/38722#discussion_r870230964)

Reply:
If only sni specified in DR, but not san. And then set auto_san_validation here, which san should it check>?

Reply:
envoy honors auto_sni and auto_san_validation config independent of each other: https://github.com/envoyproxy/envoy/blob/main/source/common/router/router.cc#L516

Reply:
I am wondering how should we disable san check when `features.VerifyCertAtClient` = true, the env indicates checking ca bundle only.

Reply:
https://istio.io/latest/docs/reference/config/networking/destination-rule/#ClientTLSSettings talks about insecureSkipVerify attribute @hzxuzhonghu

Reply:
This is two different case, i mean how to only verify ca.

Reply:
@howardjohn does it make sense to have a separate feature flag only for auto_san_validation? As per the envoy code mentioned above, both can be set independently.

Reply:
Thanks @howardjohn . In that case, probably the current PR is good enough, and you can do one more round of review?

Reply:
make sense.